Docs: Replace terms.md page with page that provides example of API syntax mapping (#3726)
ROCm-Developer-Tools/HIP: docs/reference/api_syntax.rst

.. meta::
  :description: Maps CUDA API syntax to HIP API syntax with an example
  :keywords: AMD, ROCm, HIP, CUDA, syntax, HIP syntax

********************************************************************************

> CUDA to HIP API Syntax: A Quick Comparison

Reviewer: ```suggestion
CUDA to HIP API Function Comparison
```

Reviewer: I think this content discusses the name changes from CUDA to HIP, and it should be titled accordingly.

Author: Should I also update the names in faq.rst and index.md to be inline with your suggested change?

Reviewer: That would be great. Thanks.

Author: Fixed

********************************************************************************

This page introduces key syntax differences between CUDA and HIP APIs with a focused code

> example and comparison table. For a complete list of mappings, visit :ref:`HIPIFY <HIPIFY:index>`.

Reviewer: This page provides a name map of CUDA API functions to HIP API functions, with a code sample and comparison table. 

I am not sure the code sample is needed, but it does provide some context for CUDA functions. There is no examination of any syntax changes between CUDA and HIP, which I imagine are few?

Author: Yea, the main purpose of the code sample was to provide context. Most of the syntax changes are just in the prefix, though if there are any major syntax changes, I imagine are far too few to put on display in this page

> The CUDA code block below illustrates several CUDA API syntaxes.

Reviewer: ```suggestion
The following CUDA code example illustrates several CUDA API syntaxes.
```

Author: Fixed

> .. code-block:: c

Reviewer 2: ```suggestion
.. code-block:: cpp
```

Author: Fixed

      }

      float *d_a, *d_b;
      cudaMalloc(&d_a, h_a.size() * sizeof(float));
      cudaMalloc(&d_b, h_b.size() * sizeof(float));

      cudaStream_t stream;
      cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking);

      cudaEvent_t start_event, stop_event;
      cudaEventCreate(&start_event);
      cudaEventCreate(&stop_event);

      cudaMemcpyAsync(d_a, h_a.data(), h_a.size() * sizeof(float), cudaMemcpyHostToDevice, stream);

      cudaEventRecord(start_event, stream);

      int blocks = (num_elements + threads - 1) / threads;
      block_reduction<<<blocks, threads, threads * sizeof(float), stream>>>(d_a, d_b, num_elements);

      cudaMemcpyAsync(h_b.data(), d_b, h_b.size() * sizeof(float), cudaMemcpyDeviceToHost, stream);

      cudaEventRecord(stop_event, stream);
      cudaEventSynchronize(stop_event);

      cudaEventElapsedTime(&milliseconds, start_event, stop_event);
      std::cout << "Kernel execution time: " << milliseconds << " ms\n";

      cudaFree(d_a);
      cudaFree(d_b);

      cudaEventDestroy(start_event);
      cudaEventDestroy(stop_event);
      cudaStreamDestroy(stream);

      return 0;
  }

> The table below maps CUDA API syntax to corresponding HIP API syntax, as demonstrated in the

Reviewer: ```suggestion
The following table maps CUDA API functions to corresponding HIP API functions, as demonstrated in the
```

Author: Fixed

preceding code examples.

.. list-table::
    :header-rows: 1
    :name: syntax-mapping-table

    *
      - CUDA
      - HIP

    *
      - ``#include <cuda_runtime.h>``
      - ``#include <hip/hip_runtime.h>``

    *
      - ``cudaError_t``
      - ``hipError_t``

    *
      - ``cudaEvent_t``
      - ``hipEvent_t``

    *
      - ``cudaStream_t``
      - ``hipStream_t``

    *
      - ``cudaMalloc``
      - ``hipMalloc``

    *
      - ``cudaStreamCreateWithFlags``
      - ``hipStreamCreateWithFlags``

    *
      - ``cudaStreamNonBlocking``
      - ``hipStreamNonBlocking``

    *
      - ``cudaEventCreate``
      - ``hipEventCreate``

    *
      - ``cudaMemcpyAsync``
      - ``hipMemcpyAsync``

    *
      - ``cudaMemcpyHostToDevice``
      - ``hipMemcpyHostToDevice``

    *
      - ``cudaEventRecord``
      - ``hipEventRecord``

    *
      - ``cudaEventSynchronize``
      - ``hipEventSynchronize``

    *
      - ``cudaEventElapsedTime``
      - ``hipEventElapsedTime``

    *
      - ``cudaFree``
      - ``hipFree``

    *
      - ``cudaEventDestroy``
      - ``hipEventDestroy``

    *
      - ``cudaStreamDestroy``
      - ``hipStreamDestroy``

> In summary, this comparison highlights the primary syntax differences between CUDA and HIP APIs.

Reviewer: ```suggestion
In summary, this comparison highlights the primary differences between CUDA and HIP APIs.
```

Reviewer: The name difference is not syntax per se.

Author: Fixed
